File: app/src/main/java/com/pomoremote/timer/OfflineTimer.java
package com.pomoremote.timer;

import android.os.CountDownTimer;
import com.pomoremote.service.PomodoroService;

public class OfflineTimer {
    private CountDownTimer timer;
    private final PomodoroService service;
    private TimerState state;

    public OfflineTimer(PomodoroService service) {
        this.service = service;
        this.state = new TimerState();
    }

    public void updateState(TimerState newState) {
        this.state = newState;
        if (TimerState.STATUS_RUNNING.equals(state.status)) {
            startLocalTimer();
        } else {
            stopLocalTimer();
        }
    }

    public TimerState getState() {
        return state;
    }

    private void startLocalTimer() {
        stopLocalTimer();
        
        long remainingMillis = (long) (state.remaining * 1000);
        if (remainingMillis <= 0) return;

        timer = new CountDownTimer(remainingMillis, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
                state.remaining = millisUntilFinished / 1000.0;
                service.onTimerUpdate(state);
            }

            @Override
            public void onFinish() {
                state.remaining = 0;
                state.status = TimerState.STATUS_STOPPED;
                state.completed++;
                service.onTimerComplete(state);
            }
        }.start();
    }

    private void stopLocalTimer() {
        if (timer != null) {
            timer.cancel();
            timer = null;
        }
    }
    
    public void toggle() {
        if (TimerState.STATUS_RUNNING.equals(state.status)) {
            state.status = TimerState.STATUS_PAUSED;
            state.last_action_time = System.currentTimeMillis() / 1000;
            stopLocalTimer();
            service.onTimerUpdate(state);
        } else {
            state.status = TimerState.STATUS_RUNNING;
            state.last_action_time = System.currentTimeMillis() / 1000;
            if (state.remaining <= 0) {
                state.remaining = state.duration > 0 ? state.duration : 1500;
            }
            startLocalTimer();
            service.onTimerUpdate(state);
        }
    }

    public void skip() {
        state.status = TimerState.STATUS_STOPPED;
        state.remaining = 0;
        state.last_action_time = System.currentTimeMillis() / 1000;
        stopLocalTimer();
        
        if (TimerState.PHASE_WORK.equals(state.phase)) {
            state.phase = TimerState.PHASE_SHORT;
            state.duration = 300;
            state.remaining = 300;
        } else {
            state.phase = TimerState.PHASE_WORK;
            state.duration = 1500;
            state.remaining = 1500;
        }
        service.onTimerUpdate(state);
    }
    
    public void reset() {
        state.status = TimerState.STATUS_STOPPED;
        state.remaining = state.duration;
        state.last_action_time = System.currentTimeMillis() / 1000;
        stopLocalTimer();
        service.onTimerUpdate(state);
    }
}

File: app/src/main/java/com/pomoremote/timer/OfflineTimer.kt
package com.pomoremote.timer

import android.os.CountDownTimer
import com.pomoremote.models.Session
import com.pomoremote.service.PomodoroService
import com.pomoremote.storage.HistoryRepository
import com.pomoremote.util.UtilPreferenceManager

class OfflineTimer(
    private val service: PomodoroService,
    private val prefs: UtilPreferenceManager,
    private val historyRepository: HistoryRepository
) {
    private var timer: CountDownTimer? = null
    var state: TimerState = TimerState()
        private set

    fun updateState(newState: TimerState) {
        this.state = newState
        if (TimerState.STATUS_RUNNING == state.status) {
            startLocalTimer()
        } else {
            stopLocalTimer()
        }
    }

    private fun startLocalTimer() {
        stopLocalTimer()

        val remainingMillis = (state.remaining * 1000).toLong()
        if (remainingMillis <= 0) return

        timer = object : CountDownTimer(remainingMillis, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                state.remaining = millisUntilFinished / 1000.0
                service.onTimerUpdate(state)
            }

            override fun onFinish() {
                handleTimerComplete()
            }
        }.start()
    }

    private fun stopLocalTimer() {
        timer?.cancel()
        timer = null
    }

    private fun handleTimerComplete() {
        // Record session before updating state
        val session = Session(
            type = state.phase,
            start = System.currentTimeMillis() / 1000 - state.duration.toLong(),
            duration = state.duration.toInt(),
            completed = true
        )
        historyRepository.saveSession(session)

        state.remaining = 0.0
        state.status = TimerState.STATUS_STOPPED

        if (TimerState.PHASE_WORK == state.phase) {
            state.completed++
            val longBreakAfter = prefs.longBreakAfter

            if (state.completed > 0 && state.completed % longBreakAfter == 0) {
                state.phase = TimerState.PHASE_LONG
                state.duration = (prefs.longBreakDuration * 60).toDouble()
            } else {
                state.phase = TimerState.PHASE_SHORT
                state.duration = (prefs.shortBreakDuration * 60).toDouble()
            }
        } else {
            // Break is done, back to work
            state.phase = TimerState.PHASE_WORK
            state.duration = (prefs.pomodoroDuration * 60).toDouble()
        }

        state.remaining = state.duration
        service.onTimerComplete(state)
    }

    fun toggle() {
        if (TimerState.STATUS_RUNNING == state.status) {
            state.status = TimerState.STATUS_PAUSED
            state.last_action_time = System.currentTimeMillis() / 1000
            stopLocalTimer()
            service.onTimerUpdate(state)
        } else {
            state.status = TimerState.STATUS_RUNNING
            state.last_action_time = System.currentTimeMillis() / 1000
            if (state.remaining <= 0) {
                // Should not happen usually as onFinish resets it, but just in case
                state.remaining = getDurationForPhase(state.phase)
                state.duration = state.remaining
            }
            startLocalTimer()
            service.onTimerUpdate(state)
        }
    }

    fun skip() {
        state.status = TimerState.STATUS_STOPPED
        state.last_action_time = System.currentTimeMillis() / 1000
        stopLocalTimer()

        if (TimerState.PHASE_WORK == state.phase) {
            state.phase = TimerState.PHASE_SHORT
            state.duration = (prefs.shortBreakDuration * 60).toDouble()
        } else {
            state.phase = TimerState.PHASE_WORK
            state.duration = (prefs.pomodoroDuration * 60).toDouble()
        }

        state.remaining = state.duration
        service.onTimerUpdate(state)
    }

    fun reset() {
        state.status = TimerState.STATUS_STOPPED
        state.duration = getDurationForPhase(state.phase)
        state.remaining = state.duration
        state.last_action_time = System.currentTimeMillis() / 1000
        stopLocalTimer()
        service.onTimerUpdate(state)
    }

    private fun getDurationForPhase(phase: String): Double {
        val minutes = when (phase) {
            TimerState.PHASE_WORK -> prefs.pomodoroDuration
            TimerState.PHASE_SHORT -> prefs.shortBreakDuration
            TimerState.PHASE_LONG -> prefs.longBreakDuration
            else -> 25
        }
        return (minutes * 60).toDouble()
    }
}
